General improvements

hpfr opened the pull request:
Thanks for creating this package! I made a number of fixes. Some are a bit subjective, like the change to the non-ql predicate to avoid outdated scheduled and deadline dates. Let me know what you think.

hpfr committed:
Include only open tasks by default

Previously, if a file contained solely completed tasks that happened to
have scheduled or deadline dates, those files would be included. Adding
scheduled or deadline dates to raw (non-task) headings seems
inadvisable, so prefer excluding files with those by default.

hpfr committed:
Only add agenda update hook in org-directory

Otherwise, modifying e.g. an Org task list for a software project can
add it to your agenda.

hpfr committed:
Use files that show up in any query

This is also a performance optimization: the first query that returns a
result will set `bool`, so following queries won’t be evaluated thanks
to `or` short-circuiting.

hpfr committed:
Tidy cleanup function

hpfr committed:
Use queries from org-ql-views

hpfr committed:
Improve initialization example

hpfr committed:
Use trivial org-ql-select action

hpfr committed:
Remove interactive declarations

None of these functions are intended to be called manually.

hpfr committed:
Remove file list capability from predicate

The predicate does not need to be able to operate on lists by itself; we
can call it within iteration if necessary.

hpfr committed:
Use temporary buffers instead of find-file

`find-file` selects the buffer and runs all hooks. When initializing
org-agenda-files for the first time, hundreds or more files may be
opened. `with-temp-buffer` closes the buffer after it exits, ensuring
initialization doesn’t leave open buffers for every processed Org file.
`with-current-buffer` is used for already open buffers; this is faster
and ensures no pre-existing buffers are closed during processing.

`org-ql-select` uses `find-file-noselect` under the hood. This is just
`find-file` without focusing the buffer. I am guessing it is considered
acceptable since `org-ql` does its own caching and since it can
assume you already have `org-agenda-files` set up sanely. When I
attempted to call `org-ql-select` on a buffer from `with-temp-buffer`,
it returned no results. This means initializing the `org-ql`-based
dynamic agenda will leave open all processed Org files.

hpfr committed:
Fix seq-reduce lambda argument order

Accumulated value comes first

hpfr committed:
Fix typos

nicolas-graves reviewed (commented):
I think I ensured predicates to take a list of files before separating the files between org-dynamic-agenda and org-ql-dynamic-agenda (whose call to org-ql-select can take a list of files). I don't why exactly, but I think the logic was about performance concerns when we wanted to run it on a large number of files for org-ql. Will approve this after reading other patches.

nicolas-graves commented on org-dynamic-agenda.el:
About the inclusion of only open tasks by default, I wanted to be able to include future appointements / deadlines by default, but that's indeed flawed for now. I'm going to accept this and consider that this is included in the future wanted steps.

hpfr replied:
Yeah, that would be ideal. Worth noting [from the Org manual](https://orgmode.org/manual/Deadlines-and-Scheduling.html) as well:
> Important: Scheduling an item in Org mode should not be understood in the same way that we understand scheduling a meeting. Setting a date for a meeting is just a simple appointment, you should mark this entry with a simple plain timestamp, to get this item shown on the date where it applies. This is a frequent misunderstanding by Org users. In Org mode, scheduling means setting a date when you want to start working on an action item.

So the intended use of scheduled/deadline dates is with tasks only. Ideally we would parse whether sections contain plain timestamps. Unfortunately, plain timestamps are so unstructured (they can appear anywhere in a section) that the Org Element API doesn’t include a headline property for whether one is present. For now, I limit my events to designated calendar files, but I would like to spread them across nodes as with tasks eventually too. Org QL does have a `ts` predicate, so we could probably emulate that.

nicolas-graves commented on org-dynamic-agenda.el:
```diff
@@ -86,9 +86,13 @@ optional provided FILE."
+             (file-in-directory-p (buffer-file-name) org-directory))
+    (add-hook 'before-save-hook #'org-dynamic-agenda-update-file)))
+
+(add-hook 'org-mode-hook #'org-dynamic-agenda-update-file-h)
```
What about the case where you have some org-files outside your org-directory ?

nicolas-graves replied:
That's fine for the example, but I'm not sure we should enforce this default here.

hpfr replied:
I included a bit more detail in the commit message for a37ec9dba8d3f052766ca0174453ea67a14d4b68. I don’t think we want people to be surprised by arbitrary Org files across the filesystem ending up in their agenda. The example I used in the commit message was todo lists in Emacs package source code.

I still keep my Org files in a single filesystem directory; I just happen to have enough to warrant the use of a dynamic agenda package. I guess you and others might want any Org files to appear. Personally, I think limiting to the `org-directory` is a bit safer for a package default. Maybe we could add a custom variable, something like `org-dynamic-agenda-allow-all`? But it is your package, so I’m okay with dropping this if you disagree as well!

nicolas-graves commented on org-ql-dynamic-agenda.el:
```diff
@@ -86,9 +86,9 @@ optional provided FILE or list of files."
               org-ql-dynamic-agenda-queries
-              t))
+              nil))
 
```
That would indeed be faster to confirm a file match, but will inhibit filling org-ql's cache for all queries for this cache. The best thing would be to still evaluate the cache queries but instantly return the boolean. Maybe there's a way to do this easily? By storing the result and launching a re-run of all queries in an asynchronous way before returning the result? WDYT?

hpfr replied:
To me, filling the Org QL cache seems orthogonal to this package. The first query that hits will be cached, so unless the order of the queries changes, initialization with `cleanup 'full` will still benefit from the cache. The only other time `file-p` runs is in the save hook, and since the buffer is newly modified when the save hook runs, the cache will be invalidated anyway, forcing the queries to run and strengthening the case for early matching.

If you want your Org QL queries to be fast in general, you can create your own hook running them all on buffer save, but from what I can tell that is unrelated to building the list of agenda files.

nicolas-graves replied:
Fine for me, I'll just have to reword a few things in the  README accordingly.

nicolas-graves commented on org-ql-dynamic-agenda.el:
```diff
@@ -94,13 +82,11 @@ optional provided FILE or list of files."
+                            (or file (buffer-file-name)) query)))
+              org-ql-dynamic-agenda-queries
+              t))
 
```
I think I ensured predicates to take a list of files before separating the files between org-dynamic-agenda and org-ql-dynamic-agenda (whose call to org-ql-select can take a list of files). I don't why exactly, but I think the logic was about performance concerns when we wanted to run it on a large number of files for org-ql. Will approve this after reading other patches.

hpfr replied:
Dupe from other comment, response at https://github.com/nicolas-graves/org-dynamic-agenda/pull/1#issuecomment-1710980146 :)

nicolas-graves commented on README.org:
```diff
@@ -26,11 +26,30 @@ With =org-ql= integration, =org-agenda-custom-commands= and =org-ql-views= are p
+#+end_src
 
-(org-dynamic-agenda-cleanup-files t)
+Particularly if you do not use the =org-ql= backend, you will likely want to set =org-agenda-files= to a single filename to persist =org-agenda-files= across Emacs sessions:
```
Not sure I understand why the org-ql backend should matter in this comment, can you elaborate?

hpfr replied:
IIRC I thought Org QL’s cache would avoid opening buffers for cached files, but I don’t think this is the case. Changed the wording now

nicolas-graves commented:
@hpfr Thanks a lot! I agree with most of the patch content, there's just a few points to address (maybe a bit messy, I'm not used to github reviews).

hpfr commented:
> I agree with most of the patch content, there's just a few points to address (maybe a bit messy, I'm not used to github reviews).

This reads like a review summary comment, but I don’t see any comments on the code. Did GitHub eat your review somehow?

> I think I ensured predicates to take a list of files before separating the files between org-dynamic-agenda and org-ql-dynamic-agenda (whose call to org-ql-select can take a list of files). I don't why exactly, but I think the logic was about performance concerns when we wanted to run it on a large number of files for org-ql.

If you pass `org-ql-select` a list, it [tries to find open buffers, falling back to `find-file-noselect`](https://github.com/alphapapa/org-ql/blob/898addeee956b02118f1db421a1a6c21cd5bd866/org-ql.el#L374-L379), in the same way the code I added in the non-ql backend does. I don’t think making separate calls to `org-ql-select` will have much of a performance impact versus calling it once with a list, but I haven’t benchmarked it. I guess you couldn’t have tested it either, because the arguments to the `seq-reduce` lambda were backwards.

Anyway, the predicate is called via `cl-remove-if-not`, so it only ever gets a single file in the current implementation. Definitely something we could try later, though.

nicolas-graves commented:
On 2023-09-07 19:12, Liam Hupfer wrote:

>> I agree with most of the patch content, there's just a few points to address (maybe a bit messy, I'm not used to github reviews).
>
> This reads like a review summary comment, but I don’t see any comments on the code. Did GitHub eat your review somehow?

You should be able to see most of my comments here
https://github.com/nicolas-graves/org-dynamic-agenda/pull/1

At least I do

There are 4-5 comments ;) 

>
>> I think I ensured predicates to take a list of files before separating the files between org-dynamic-agenda and org-ql-dynamic-agenda (whose call to org-ql-select can take a list of files). I don't why exactly, but I think the logic was about performance concerns when we wanted to run it on a large number of files for org-ql.
>
> If you pass `org-ql-select` a list, it [tries to find open buffers, falling back to `find-file-noselect`](https://github.com/alphapapa/org-ql/blob/898addeee956b02118f1db421a1a6c21cd5bd866/org-ql.el#L374-L379), in the same way the code I added in the non-ql backend does. I don’t think making separate calls to `org-ql-select` will have much of a performance impact versus calling it once with a list, but I haven’t benchmarked it. I guess you couldn’t have tested it either, because the arguments to the `seq-reduce` lambda were backwards.

Thanks for this fix by the way
>
> Anyway, the predicate is called via `cl-remove-if-not`, so it only ever gets a single file in the current implementation. Definitely something we could try later, though.

Fine for me, it's versionned so we'll still be able to do it later yes. 

-- 
Best regards,
Nicolas Graves

hpfr commented:
> You should be able to see most of my comments here #1
> At least I do
> There are 4-5 comments ;)

https://github.com/doomemacs/doomemacs/pull/7260 is a PR with review comments that I can see. The “Conversation” tab says `11` comments are present, and if I expand the resolved threads, I do count 11 comments total. In this PR (#1), with the addition of this comment, the “Conversation” tab will say `5` comments are present: two from you, then my previous reply, then your last reply, then this comment. If you can see them, can you use the `…` button to copy a link for me? Or grab a screenshot maybe? I am confused :sweat_smile:.

nicolas-graves commented:
On 2023-09-08 15:26, Liam Hupfer wrote:

>> You should be able to see most of my comments here #1
>> At least I do
>> There are 4-5 comments ;)
>
> https://github.com/doomemacs/doomemacs/pull/7260 is a PR with review comments that I can see. The “Conversation” tab says `11` comments are present, and if I expand the resolved threads, I do count 11 comments total. In this PR (#1), with the addition of this comment, the “Conversation” tab will say `5` comments are present: two from you, then my previous reply, then your last reply, then this comment. If you can see them, can you use the `…` button to copy a link for me? Or grab a screenshot maybe? I am confused :sweat_smile:.

My bad, my review was saved but not submitted. Should be accessible now,
sorry again this is one of the first times I use this workflow. 

-- 
Best regards,
Nicolas Graves

hpfr committed:
Add save hook buffer-locally

Before, the file update function would run for all buffers once added to
the hook.

hpfr committed:
Fix removal case in update function

Unlike `cl-pushnew`, `cl-delete` does not operate on generalized
variables.

hpfr committed:
Remove org-ql queries cache

Parsing these two variables seems to take very little time compared to
running the actual queries. This ensures that when the user changes
their `org-agenda-custom-commands` or `org-ql-views`, their dynamic
agenda will immediately work according to the new queries.

hpfr reviewed (commented):
In the process of testing to respond to your review I found some more fixes :sweat_smile: I think I finally have the Org QL backend working fairly well.

Edit: evidently I am new to the GitHub workflow as well. Unlike GitLab, if you “start a review” to stage replies locally, it renders them all twice.
